### PR TITLE
Enable out-of-tree projects to use MLIR.

### PR DESCRIPTION
**Pull request by sscalpone**
Add CMake rules to copy various tools and files into the install directory.  Installs the include and td sources, as well as generated inc files, in include/mlir.  Installs the binaries mlir-cpu-runner, mlir-cuda-runner, mlir-opt, mlir-translate, and mlir-tblgen in bin.

### CODE REVIEW

File: tools/mlir-cpu-runner/CMakeLists.txt [hunk 1]
```diff
@@ -21,3 +21,14 @@ target_link_libraries(mlir-cpu-runner PRIVATE
...
 )
+
+install(TARGETS mlir-cpu-runner
+  ${export_to_llvmexports}
+  COMPONENT mlir-cpu-runner
+  RUNTIME DESTINATION ${LLVM_TOOLS_INSTALL_DIR})
+
+if(NOT LLVM_ENABLE_IDE)
+  add_llvm_install_targets("install-mlir-cpu-runner"
+    DEPENDS mlir-cpu-runner
```
Comment (schweitzpgi) [inline tools/mlir-cpu-runner/CMakeLists.txt:32]: Do we need mlir-cpu-runner?

File: tools/mlir-cuda-runner/CMakeLists.txt [hunk 2]
```diff
@@ -71,4 +71,15 @@ if(MLIR_CUDA_RUNNER_ENABLED)
...
   target_link_libraries(mlir-cuda-runner PRIVATE ${FULL_LINK_LIBS} ${LIBS})
 
+  install(TARGETS mlir-cuda-runner
+    ${export_to_llvmexports}
+    COMPONENT mlir-cuda-runner
+    RUNTIME DESTINATION ${LLVM_TOOLS_INSTALL_DIR})
+
+  if(NOT LLVM_ENABLE_IDE)
+    add_llvm_install_targets("install-mlir-cuda-runner"
+      DEPENDS mlir-cuda-runner
```
Comment (schweitzpgi) [inline tools/mlir-cuda-runner/CMakeLists.txt:81]: Does f18 need mlir-cuda-runner?

**Review by schweitzpgi — COMMENTED** [review]
This seems to me like a largish number of files and diffs to perform what should be a straightforward installation config.  I'll try to prune this down and post a couple of PRs.

File: tools/mlir-cpu-runner/CMakeLists.txt (hunk 1 above)
Comment (sscalpone) [reply tools/mlir-cpu-runner/CMakeLists.txt:32]: I focused on directories with files needed by f18.  I didn't try to cull the number of files from each of those directory to match exactly the needs of f18.  I.e. I was trying to be complete for each directory that I touched.

**Comment by sscalpone** [comment]
The mlir push got a couple good suggestions for changes.

**Comment by schweitzpgi** [comment]
Take a look at https://github.com/flang-compiler/f18-mlir/pull/3 as well.

**Pull request closed without merging**